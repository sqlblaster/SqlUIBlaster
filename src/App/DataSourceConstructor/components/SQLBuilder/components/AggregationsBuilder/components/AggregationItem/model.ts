import { Column } from 'src/App/DataSourceConstructor/schemas';
import * as uuid from 'uuid';
import { ColumnBranch } from '../../../common/ColumnSelector/ColumnItem/SelectedColumn.models';
import { AggregationOperator } from '../../aggregations';

export class Aggregation {

  public id: string;
  public operator: AggregationOperator;
  public column: ColumnBranch | undefined;
  public columnFilterTypes: ColumnFilterType[];

  constructor(...inits: Partial<Aggregation>[]) {
    Object.assign(this, ...inits);

    if (!this.id) {
      this.id = uuid();
    }

    this.columnFilterTypes = Aggregation.getExpectedColumnsTypeFrom(
      this.operator
    );
    !this.columnFilterTypes.length && (this.column = undefined);
  }

  public static hasNoColumn(operator: AggregationOperator): boolean {
    return (
      operator === 'Count of rows' || operator === 'Cumulative count of rows'
    );
  }

  public static getExpectedColumnsTypeFrom(
    operator: AggregationOperator
  ): Aggregation['columnFilterTypes'] {
    if (operator === 'Number of distinct values of') {
      return ['all'];
    } else if (!Aggregation.hasNoColumn(operator)) {
      return ['number'];
    } else {
      return [];
    }
  }
}

export type ColumnFilterType = Column['type'] | 'all' | undefined;
