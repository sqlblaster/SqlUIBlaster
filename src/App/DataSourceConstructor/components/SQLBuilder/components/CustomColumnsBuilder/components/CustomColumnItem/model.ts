import * as uuid from 'uuid';
import { ColumnBranch } from '../../../common/ColumnSelector/ColumnItem/SelectedColumn.models';

export class CustomColumn {
  public id: string;
  public column: ColumnBranch;

  constructor(...inits: Partial<CustomColumn>[]) {
    Object.assign(this, ...inits);

    if (!this.id) {
      this.id = uuid();
    }
  }
}
