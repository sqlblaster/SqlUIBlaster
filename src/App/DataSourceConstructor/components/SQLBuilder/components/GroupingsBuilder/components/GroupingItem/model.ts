import uuid from 'uuid';
import { ColumnBranch } from '../../../common/ColumnSelector/ColumnItem/SelectedColumn.models';

export class Grouping {
  public id: string;
  // TODO: rename column to columnBranch
  public column: ColumnBranch;

  constructor(...inits: Partial<Grouping>[]) {
    Object.assign(this, ...inits);

    if (!this.id) {
      this.id = uuid();
    }
  }
}
