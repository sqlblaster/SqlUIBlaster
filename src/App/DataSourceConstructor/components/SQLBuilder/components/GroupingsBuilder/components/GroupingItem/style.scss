.grouping-item {
  display: flex;
  align-items: center;
  border: 2px solid transparent;
  border-radius: 3px;
  transition: 0.3s;
  margin: 5px 0px;
}
