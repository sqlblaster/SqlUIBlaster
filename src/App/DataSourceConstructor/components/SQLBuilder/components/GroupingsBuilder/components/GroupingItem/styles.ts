import { StyleRulesCallback } from '@material-ui/core';

export type GroupingItemClassKeys = 'root' | 'column-view';

export const groupingItemStyles: StyleRulesCallback<
  GroupingItemClassKeys
> = (theme) => ({
  root: {
    display: 'flex',
    alignItems: 'center',
    padding: 2,
    borderRadius: '3px',
    transition: '0.3s',
    margin: '5px 0px'
  },
  'column-view': {
    cursor: 'pointer',
    '&:hover': {
      opacity: 0.8
    },
    transition: '0.3s'
  }
});
