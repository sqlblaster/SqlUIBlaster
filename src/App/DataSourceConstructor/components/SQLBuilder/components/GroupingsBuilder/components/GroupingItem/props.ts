import { ClassesProp } from 'src/App/utils/classes-prop';
import { actionCreators, PickActionCreators } from 'src/store/action-creators';
import { PopoverManagerProps } from '../../../common/WithPopoverManagement';
import { Grouping } from './model';
import { GroupingItemClassKeys } from './styles';

export interface GroupingItemOwnProps {
  grouping: Grouping;
}

export type GroupingItemDispatchProps = PickActionCreators<
  'removeGrouping' | 'updateGrouping'
>;

export const GIMapDispatchToProps: GroupingItemDispatchProps = {
  removeGrouping: actionCreators.removeGrouping,
  updateGrouping: actionCreators.updateGrouping
};

export type GroupingItemProps = GroupingItemOwnProps &
  GroupingItemDispatchProps;

export type GroupingItemViewProps = {
  handleGroupingRemoval: (id: Grouping['id']) => () => void;
} & GroupingItemOwnProps &
  PopoverManagerProps &
  PickActionCreators<'updateGrouping'> &
  ClassesProp<GroupingItemClassKeys>;
