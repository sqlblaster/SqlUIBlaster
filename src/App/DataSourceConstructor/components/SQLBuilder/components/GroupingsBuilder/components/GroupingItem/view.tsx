import { withStyles } from '@material-ui/core';
import * as React from 'react';
import { RemoveButton } from 'src/App/DataSourceConstructor/components/SQLBuilder/components/common/Icons/RemoveButton';
import { SequentialColumnView } from 'src/App/DataSourceConstructor/components/SQLBuilder/components/common/SequentialColumnView';
import { MutateGroupingPopover } from '../MutateGroupingPopover';
import { GroupingItemViewProps } from './props';
import { groupingItemStyles } from './styles';

export const GroupingItem = withStyles(groupingItemStyles)(
  class extends React.Component<GroupingItemViewProps> {
    private readonly popoverLauncher: React.RefObject<HTMLDivElement>;

    constructor(props: GroupingItemViewProps) {
      super(props);

      this.popoverLauncher = React.createRef<HTMLDivElement>();
    }

    public render() {
      const {
        grouping,
        updateGrouping,
        classes,
        isOpen,
        handleOpen,
        handleClose,
        handleGroupingRemoval
      } = this.props;

      const { id, column } = grouping;

      return (
        <>
          <div className={classes.root}>
            <div
              ref={this.popoverLauncher}
              onClick={handleOpen}
              className={classes['column-view']}
              data-testid='column-view'
            >
              <SequentialColumnView {...column} />
            </div>
            <RemoveButton
              onClick={handleGroupingRemoval(id)}
              data-testid='remove'
            />
          </div>
          { isOpen && (
            <MutateGroupingPopover
              isOpen={isOpen}
              anchorEl={this.popoverLauncher.current}
              action={updateGrouping}
              onClose={handleClose}
              grouping={grouping}
            />
          ) }
        </>
      );
    }
  }
);
