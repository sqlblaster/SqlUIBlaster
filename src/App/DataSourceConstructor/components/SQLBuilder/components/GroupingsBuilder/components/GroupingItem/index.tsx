import * as React from 'react';
import { connect } from 'react-redux';
import { State } from 'src/store/models/State';
import { WithPopoverManagement } from '../../../common/WithPopoverManagement';
import {
  GIMapDispatchToProps,
  GroupingItemDispatchProps,
  GroupingItemOwnProps,
  GroupingItemProps,
  GroupingItemViewProps as ViewProps
} from './props';
import { GroupingItem as View } from './view';

export const GroupingItem = connect<
  {},
  GroupingItemDispatchProps,
  GroupingItemOwnProps,
  State
>(
  null,
  GIMapDispatchToProps
)(
  class extends React.Component<GroupingItemProps> {
    public static ViewWithPopoverManagement = WithPopoverManagement(View);

    public handleGroupingRemoval: ViewProps['handleGroupingRemoval'] = (id) => () =>
      this.props.removeGrouping(id)

    public render() {
      return (
        <GroupingItem.ViewWithPopoverManagement
          handleGroupingRemoval={this.handleGroupingRemoval}
          {...this.props}
        />
      );
    }
  }
);
