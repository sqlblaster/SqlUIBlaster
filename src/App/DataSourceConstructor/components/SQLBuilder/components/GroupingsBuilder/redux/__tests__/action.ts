import { Grouping } from '../../components/GroupingItem/model';
import {
  addGrouping,
  clearGroupings,
  removeGrouping,
  updateGrouping
} from '../action';

describe('GroupingsBuilder action creators should return appropriate actions', () => {
  it('addGrouping', () => {
    expect(
      addGrouping(new Grouping({ id: '566af397-c08b-42b7-b10b-a4548e0a17ba' }))
    ).toMatchSnapshot();
  });

  it('updateGrouping', () => {
    expect(
      updateGrouping(
        new Grouping({ id: 'c3c15257-27e4-49bb-8803-4c9d87e7cd77' })
      )
    ).toMatchSnapshot();
  });

  it('removeGrouping', () => {
    const id = '1';
    expect(removeGrouping(id)).toMatchSnapshot();
  });

  it('clearGroupings', () => {
    expect(clearGroupings()).toMatchSnapshot();
  });
});
