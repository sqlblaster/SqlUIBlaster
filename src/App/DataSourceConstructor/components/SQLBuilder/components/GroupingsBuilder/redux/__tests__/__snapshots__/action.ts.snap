// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GroupingsBuilder action creators should return appropriate actions addGrouping 1`] = `
Object {
  "grouping": Grouping {
    "id": "566af397-c08b-42b7-b10b-a4548e0a17ba",
  },
  "type": "ADD_GROUPING",
}
`;

exports[`GroupingsBuilder action creators should return appropriate actions clearGroupings 1`] = `
Object {
  "type": "CLEAR_GROUPINGS",
}
`;

exports[`GroupingsBuilder action creators should return appropriate actions removeGrouping 1`] = `
Object {
  "id": "1",
  "type": "REMOVE_GROUPING",
}
`;

exports[`GroupingsBuilder action creators should return appropriate actions updateGrouping 1`] = `
Object {
  "grouping": Grouping {
    "id": "c3c15257-27e4-49bb-8803-4c9d87e7cd77",
  },
  "type": "UPDATE_GROUPING",
}
`;
